deepreg/model/layer.py

```diff
@@ -425,25 +427,31 @@ def __init__(
             0.44 * cp for cp in control_point_spacing
         ]  # 0.44 = ln(4)/pi
         self.cp_spacing = control_point_spacing
-        self.kernel = None
+        self.kernel = [gaussian_kernel1d(sigma=sigma) for sigma in self.kernel_sigma]
         self._output_shape = None
         self._resize = None
 
     def build(self, input_shape):
         super().build(input_shape=input_shape)
 
-        self.kernel = layer_util.gaussian_filter_3d(self.kernel_sigma)
         output_shape = tuple(
-            tf.cast(tf.math.ceil(v / c) + 3, tf.int32)
-            for v, c in zip(input_shape[1:-1], self.cp_spacing)
+            [
+                tf.cast(tf.math.ceil(v / c) + 3, tf.int32)
+                for v, c in zip(input_shape[1:-1], self.cp_spacing)
+            ]
         )
         self._output_shape = output_shape
         self._resize = Resize3d(output_shape)
 
     def call(self, inputs, **kwargs) -> tf.Tensor:
-        output = tf.nn.conv3d(

```

Comment:
I thought we just need to replace `conv3d` by `separable_filter` ? why you got an extra for loop?

Reply:
Our current `separable_filter` implementation assumes that channel dimension = 1. Since the filter has to be applied independently to each channel, I use the for loop to apply the filter to each channel.

Reply:
I see. Let me have another look later :)